mlir/lib/Quantum/Transforms/ChainedSelfInversePatterns.cpp

```diff
@@ -30,24 +30,32 @@ namespace {
 struct ChainedHadamardOpRewritePattern : public mlir::OpRewritePattern<CustomOp> {
     using mlir::OpRewritePattern<CustomOp>::OpRewritePattern;
 
-    /// We check if the operation and it's parent are hadamard operations. If so, replace op
-    /// with it's "grandparent".
+    /// We simplify consecutive Hermitian quantum gates and simplifies them.
+    /// Hermitian gates are self-inverse and applying the same gate twice in succession
+    /// cancels out the effect. This pattern rewrites such redundant operations by
+    /// replacing the operation with its "grandparent" operation in the quantum circuit.
     mlir::LogicalResult matchAndRewrite(CustomOp op, mlir::PatternRewriter &rewriter) const override
     {
-        LLVM_DEBUG(dbgs() << "Simplifying the following hadamard operation:\n" << op << "\n");
-        if (op.getGateName().str() != "Hadamard")
-            return failure();
-
-        ValueRange qbs = op.getInQubits();
-        auto parentHadamard = dyn_cast<CustomOp>(qbs[0].getDefiningOp());
+        LLVM_DEBUG(dbgs() << "Simplifying the following operation:\n" << op << "\n");
 
-        if (parentHadamard == nullptr)
+        std::set<StringRef> HermitianOps{"Hadamard", "PauliX", "PauliY", "PauliZ", "CNOT",
```

Comment:
I think llvm has a container type for a set of strings https://llvm.org/docs/ProgrammersManual.html#llvm-adt-stringset-h